### ISSUE TITLE
Solr crashes on JDK17+ during JIT compile of BoundedLocalCache::put

### ISSUE THREAD

**Issue by janhoy** [ben-manes/caffeine#797]
Solr uses caffeine, and several users have reported crashes during JIT compilation of Caffeine. See https://issues.apache.org/jira/browse/SOLR-16463 for our tracking of this.

This is likely a JDK bug (see https://bugs.openjdk.org/browse/JDK-8285835) but wanted to highlight it over here too, in case you find a way to reproduce or circumvent it.

**Comment by janhoy** [comment]
We are experimenting with a workaround for Solr, excluding JIT optimization of this method, see https://github.com/apache/solr-docker/pull/11

**Comment by uschindler** [comment]
This should be the workaround (Java command line arg at runtime): `-XX:CompileCommand=exclude,com.github.benmanes.caffeine.cache.BoundedLocalCache::put`

**Comment by uschindler** [comment]
I agree there's nothing to do. You may only reproduce it with heavy testing for hours. The jfk people now found a way to trigger the issue in Lucene with special command line parameters.

But for sure the loop on that method is somehow special and causes the bug due to loop unrolling. My guess: it unrolls the loop as it mostly only runs exactly one time. In the seldom case the loop runs twice, but as code is optimized the compiler gets into illegal state and runs into the assertion.

**Comment by uschindler** [comment]
Hi,
there is now a PR: https://github.com/openjdk/jdk/pull/10894

@ben-manes : could you check the loop's code if you find a similar pattern in there? For Lucene I figured out how it happens. Basically if you have a loop accessing object instances that refer to each other in three layers. A wraps/references B wraps/references C. I don't undertsnad your code, but to me it looks similar (there are many nodes created and they refer to each other).

**Comment by ben-manes** [comment]
I suppose it might. The node holds the key and value, where the key might be abstracted with an intermediate object if using a weak key reference caching. As the map wraps it in its own entry object, I could see a few layers. The LRU doubly-linked lists are handled elsewhere, so that type of relationship is not manipulated here. It is likely the same issue that this change fixes, but I cannot be certain from comparing Caffeine's code to the test cases.

**Comment by uschindler** [comment]
The only way to check this is to wait for all Backports done and released and then let people try.
